lib/srv/multireader.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"io"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// MultiReader is a wrapper around io.MultiReader which allows to add and remove readers
+type MultiReader struct {
+	sync.RWMutex
+	readers map[string]io.Reader
+	multi   io.Reader
+}
+
+// NewMultiReader creates a new multitreader.
+func NewMultiReader() *MultiReader {
+	return &MultiReader{
+		readers: make(map[string]io.Reader),
+		multi:   io.MultiReader(),
+	}
+}
+
+// Read reads from any of the underlying readers.
+func (r *MultiReader) Read(p []byte) (int, error) {
+	r.RLock()
+	multi := r.multi
+	r.RUnlock()
+
+	read, err := multi.Read(p)
+	if err != nil {
+		log.Warnf("failed to read from multireader: %v", err)
+	}
+
+	return read, nil
+}
+
+// AddReader adds the reader.
+func (r *MultiReader) AddReader(name string, reader io.Reader) {
```

Comment:
What is the purpose of this?

`io.MultiReader` is meant to concatenate readers, so _order matters_. It reads from them in order and when one reader is fully consumed, it moves to the next one.

This is constructing a multireader based on the result of a map iteration, which has no defined order. This means every time you add a reader, you're effectively shuffling the readers and restarting from the beginning.

Is this what you want? It doesn't seem correct.

Reply:
I've completely revamped how multiplexed io is done which cleans everything up.